src/framework/util/async_mutex.ts

```diff
@@ -0,0 +1,21 @@
+export class AsyncMutex {
+  private lastToRun: Promise<unknown> | undefined;
+
+  // Run an async function with a lock on this mutex.
+  // Waits until the mutex is available, locks it, runs the function, then releases it.
+  async with<T>(fn: () => Promise<T>): Promise<T> {
+    const p = (async () => {
+      // If the mutex is locked, wait for the last thing in the queue before running.
```

Comment:
nit: Now the queue is sort of implicitly defined by a chain of promises. These comments may be a bit confusing.

Reply:
Done